File: decrypt-messages.py
#!/usr/bin/env python3
"""
Standalone decryption utility for FTJM Chat messages

Requirements:
  pip install pycryptodome

Usage:
  python3 decrypt-messages.py "gc:U2FsdGVkX18..."
  python3 decrypt-messages.py (interactive mode)
"""

import sys
import hashlib
from Crypto.Cipher import AES
from Crypto.Util.Padding import unpad

SECRET_KEY = 'your-secret-encryption-key-at-least-32-chars'
GC_PREFIX = 'gc:'


def evp_bytes_to_key(password: str, salt: bytes, key_len: int, iv_len: int) -> tuple:
    """
    Derive key and IV using OpenSSL's EVP_BytesToKey
    Compatible with CryptoJS AES encryption
    """
    m = []
    i = 0
    while len(b''.join(m)) < (key_len + iv_len):
        md5 = hashlib.md5()
        data = password.encode() if i == 0 else m[i - 1] + password.encode()
        md5.update(data + salt)
        m.append(md5.digest())
        i += 1
    ms = b''.join(m)
    return ms[:key_len], ms[key_len : key_len + iv_len]


def decrypt_text(ciphertext: str) -> str | None:
    """Decrypt CryptoJS AES encrypted text"""
    if not ciphertext:
        return ciphertext

    try:
        # Remove prefix
        payload = (
            ciphertext[len(GC_PREFIX) :]
            if ciphertext.startswith(GC_PREFIX)
            else ciphertext
        )

        # Decode base64
        import base64

        encrypted_bytes = base64.b64decode(payload)

        # Extract salt (after "Salted__" magic bytes)
        if not encrypted_bytes.startswith(b"Salted__"):
            raise ValueError("Invalid encrypted format: missing Salted__ header")

        salt = encrypted_bytes[8:16]
        encrypted = encrypted_bytes[16:]

        # Derive key and IV
        key, iv = evp_bytes_to_key(SECRET_KEY, salt, 32, 16)

        # Decrypt using AES-256-CBC
        cipher = AES.new(key, AES.MODE_CBC, iv)
        decrypted = cipher.decrypt(encrypted)

        # Remove PKCS#7 padding
        unpadded = unpad(decrypted, AES.block_size)

        return unpadded.decode("utf-8")

    except Exception as e:
        print(f"❌ Decryption failed: {e}")
        return None


def interactive_mode():
    """Interactive decryption mode"""
    print("🔐 FTJM Chat Message Decryption Utility")
    print("======================================")
    print('Paste encrypted messages (starting with "gc:") one per line.')
    print('Type "quit" or "exit" to stop.\n')

    while True:
        try:
            user_input = input("Enter encrypted message (or 'quit'): ").strip()

            if not user_input or user_input.lower() in ("quit", "exit"):
                print("Goodbye! 👋")
                break

            decrypted = decrypt_text(user_input)
            if decrypted:
                print(f'✅ Decrypted: "{decrypted}"\n')
            else:
                print("⚠️  Could not decrypt message\n")

        except KeyboardInterrupt:
            print("\nGoodbye! 👋")
            break


def main():
    if len(sys.argv) > 1:
        # Command line mode
        encrypted = " ".join(sys.argv[1:])
        print(f"🔐 Decrypting: {encrypted[:50]}...")

        decrypted = decrypt_text(encrypted)
        if decrypted:
            print(f'✅ Result: "{decrypted}"')
            sys.exit(0)
        else:
            sys.exit(1)
    else:
        # Interactive mode
        interactive_mode()


if __name__ == "__main__":
    main()

File: decrypt-messages.js
#!/usr/bin/env node

/**
 * Standalone decryption utility for FTJM Chat messages
 * 
 * Usage:
 *   node decrypt-messages.js "gc:U2FsdGVkX18..."
 *   node decrypt-messages.js (interactive mode)
 */

import crypto from 'crypto';
import readline from 'readline';

const SECRET_KEY = 'app-chat-secret-key-2024';
const GC_PREFIX = 'gc:';

/**
 * Decrypt CryptoJS AES encrypted text
 * Supports OpenSSL EVP_BytesToKey format (default for CryptoJS)
 */
function decryptText(ciphertext) {
  if (!ciphertext) return ciphertext;

  try {
    // Remove prefix
    const payload = ciphertext.startsWith(GC_PREFIX)
      ? ciphertext.slice(GC_PREFIX.length)
      : ciphertext;

    // Decode base64
    const encryptedBytes = Buffer.from(payload, 'base64');

    // Extract salt (after "Salted__" magic bytes)
    if (!encryptedBytes.toString('utf8', 0, 8).startsWith('Salted__')) {
      throw new Error('Invalid encrypted format: missing Salted__ header');
    }

    const salt = encryptedBytes.slice(8, 16);
    const encrypted = encryptedBytes.slice(16);

    // Derive key and IV using OpenSSL EVP_BytesToKey
    // CryptoJS uses: MD5(password + salt) repeatedly to derive key + iv
    const md5 = crypto.createHash('md5');
    md5.update(SECRET_KEY + salt.toString('binary'));
    const m_i = md5.digest();

    const md5_2 = crypto.createHash('md5');
    md5_2.update(m_i + SECRET_KEY + salt.toString('binary'));
    const m_ii = md5_2.digest();

    const key = Buffer.concat([m_i, m_ii]);
    const iv = key.slice(32, 48);

    // Decrypt using AES-256-CBC
    const decipher = crypto.createDecipheriv('aes-256-cbc', key.slice(0, 32), iv);
    let decrypted = decipher.update(encrypted);
    decrypted = Buffer.concat([decrypted, decipher.final()]);

    // Remove PKCS#7 padding
    const paddingLength = decrypted[decrypted.length - 1];
    const unpadded = decrypted.slice(0, decrypted.length - paddingLength);

    return unpadded.toString('utf8');
  } catch (error) {
    console.error(`❌ Decryption failed: ${error.message}`);
    return null;
  }
}

/**
 * Interactive mode
 */
async function interactiveMode() {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  console.log('🔐 FTJM Chat Message Decryption Utility');
  console.log('======================================');
  console.log('Paste encrypted messages (starting with "gc:") one per line.');
  console.log('Type "quit" or "exit" to stop.\n');

  const askForInput = () => {
    rl.question('Enter encrypted message (or "quit"): ', (input) => {
      if (!input || input.toLowerCase() === 'quit' || input.toLowerCase() === 'exit') {
        console.log('Goodbye! 👋');
        rl.close();
        return;
      }

      const decrypted = decryptText(input);
      if (decrypted) {
        console.log(`✅ Decrypted: "${decrypted}"\n`);
      } else {
        console.log('⚠️  Could not decrypt message\n');
      }

      askForInput();
    });
  };

  askForInput();
}

/**
 * Main
 */
if (process.argv.length > 2) {
  // Command line mode
  const encrypted = process.argv.slice(2).join(' ');
  console.log(`🔐 Decrypting: ${encrypted.substring(0, 50)}...`);

  const decrypted = decryptText(encrypted);
  if (decrypted) {
    console.log(`✅ Result: "${decrypted}"`);
    process.exit(0);
  } else {
    process.exit(1);
  }
} else {
  // Interactive mode
  interactiveMode();
}
